src/protocols/bdx/BDXMessageUtils.h

```diff
@@ -0,0 +1,231 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines structures and utility methods for working with BDX
+ *      messages.
+ */
+
+#pragma once
+
+#include <support/CodeUtils.h>
+#include <system/SystemPacketBuffer.h>
+
+namespace chip {
+namespace BDX {
+
+enum ControlMode : uint8_t
+{
+    kNotSpecified  = 0x00,
+    kSenderDrive   = 0x10,
+    kReceiverDrive = 0x20,
+    kAsync         = 0x40,
+};
+
+/*
+ * A structure for representing a SendInit or ReceiveInit message (both contain
+ * identical parameters).
+ */
+struct TransferInit
+{
+    /**
+     * @brief
+     *  Pack (write) the message into a PacketBuffer.
+     *
+     * @param[out] aBuffer a PacketBuffer to use to write the message
+     *
+     * @return CHIP_ERROR Any error that occurs when trying to write to the PacketBuffer
+     */
+    CHIP_ERROR Pack(System::PacketBuffer & aBuffer);
+
+    /**
+     * @brief
+     *  Parse data from an PacketBuffer into a struct instance
+     *
+     * @param[in] aBuffer Pointer to a PacketBuffer containing the data.
+     * @param[out] aParsedMessage Reference to a struct instance where results will be stored.
+     *             Note that this struct will store pointers into the passed PacketBuffer, so it
+     *             is essential that the PacketBuffer is not modified or freed until after the
+     *             TransferInit struct is no longer needed.
+     *
+     * @return CHIP_ERROR Any error that occurs when trying to read the message
+     */
+    static CHIP_ERROR Parse(System::PacketBuffer & aBuffer, TransferInit & aParsedMessage);
+
+    /**
+     * @brief
+     *  Get the size of the message once it is written to a buffer.
```

Comment:
size necessary to pack into a buffer?

the 'once it is written' sounds as if you need to write it first.

Reply:
Good point, I'll change the wording